### PR TITLE
Nested Scrollview fix

### PR DESCRIPTION
**Pull request by Protothor**
Solution to issue #164 
Gets first scrollView in Element.scroll() to fix iOS nested scrollView problem

### CODE REVIEW

**Comment by rotemmiz** [comment]
Can you please provide an e2e test that will prove this correction?

**Comment by LeoNatan** [comment]
I don't think this is a correct fix. As I understand from the code, this will only scroll the first matched instance. But as the issue describes, if two scroll views exist, outer for vertical and inner for horizontal scrolls, scrolling vertically with this solution would fail.

This is a good workaround for people that need the type of solution here, but this is not a general case.

Correct me if I'm wrong.

**Comment by Protothor** [comment]
Sorry about not getting back to you until now, but I live on the West Coast of the US (GMT -8)

Based on my testing it works, and in theory based on the code it should too. The "_extendToDescendantScrollViews" function gets all UIScrollViews inside of the RCTScrollView and the order they will appear in array will be from parent to child. Since the first one is always the first UIScrollView that is returned from that array will be the one we're targeting based on a testID the "atIndex(0)" will target the correct one. Unfortunately, I didn't have the time to attack the "waitFor().action().whileElement().scroll()" this time because it uses a great deal of native code to do the whileElement matcher, but this allows testing of nested ScrollViews in iOS in some fashion even if it's not as pretty.

**Comment by Protothor** [comment]
So, I'm not sure what the problem with the 4th build.

The first time it froze on the "before all" hook
>   1) "before all" hook
>   0 passing (2m)
>   1 failing
>   1) "before all" hook:
>      Error: Timeout of 120000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.

And the second time it timed out during the lerna set up
> $ ./scripts/demo-projects.ios.sh
>  ./scripts/bootstrap.sh 
> lerna info version 2.4.0
> lerna info Bootstrapping 9 packages
> lerna info lifecycle preinstall
> lerna info Installing external dependencies
> No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
> Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
> The build has been terminated

Otherwise the tests works, it's a simple nested scrollview that uses "swipe", "scrollTo" and "scroll" to scroll the both the parent and child scroll views.

**Comment by Protothor** [comment]
Hello again, I didn't want to bother anyone around the end of the year holidays, but this passes, there are tests written, is there anything else I can do to complete this pull request?

**Comment by rotemmiz** [comment]
Looks great, I just need you to make sure it doesn't break Android e2e, and since we still can't run it in CI please make sure[ it runs locally ](https://github.com/wix/detox/blob/master/docs/Guide.Contributing.md#android). When done mention it here and I'll merge

**Comment by Protothor** [comment]
So, I tested android and the way that the deep linking worked there, it was getting stuck. So I just updated the test to launch a non-linked instance so that it wouldn't lock the app in after a device.reloadReactNative()

**Comment by Protothor** [comment]
Hello, checking in again. Seeing where we're at with this. It's all passing. I've verified android locally. Anything else I can do to get this merged?

**Comment by lorrainefox** [comment]
This update is very important to my project. Can we get an approval please?  I have some people on my team looking for another automated testing tool and I would rather stay with detox.

**Comment by rotemmiz** [comment]
The best way to to have this PR merged is by adding an e2e test that proves its 
necessity. This e2e test will also make sure it doesn't break in the future.
This is where I'd add it:
E2e test:
https://github.com/wix/detox/blob/master/detox/test/e2e/c-actions.js

Test app:
https://github.com/wix/detox/blob/master/detox/test/src/Screens/ActionsScreen.js

**Comment by Protothor** [comment]
Oh, I'm sorry if I placed the tests incorrectly. I added another test suite. (p-nested-scroll-view), linked it to the main page and ran it as a part of the suite.

File: detox/test/e2e/n-deep-links.js [hunk 1]
```diff
@@ -1,4 +1,7 @@
 describe('Deep Links', () => {
+  after(async () => {
```
Comment (rotemmiz) [inline detox/test/e2e/n-deep-links.js]: Why was this added?

File: detox/test/e2e/p-nested-scroll-views.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
+describe('Nested Scroll Views', () => {
+  beforeEach(async () => {
+    await device.reloadReactNative();
+  });
+
+  beforeEach(async () => {
```
Comment (rotemmiz) [inline detox/test/e2e/p-nested-scroll-views.js]: You can merge both` beforeEach`s

File: detox/test/e2e/p-nested-scroll-views.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    await element(by.text('Nested ScrollView')).tap();
+  });
+
+  it('Should handle swipe', async () => {
```
Comment (rotemmiz) [inline detox/test/e2e/p-nested-scroll-views.js]: Nicee!!

File: detox/test/e2e/p-nested-scroll-views.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+    await element(by.id('childScrollView')).scrollTo('right');
+    await expect(element(by.text('Right'))).toBeVisible();
+    await expect(element(by.text('Left'))).toBeNotVisible();
+    await element(by.id('childScrollView')).scrollTo('left');
+    await expect(element(by.text('Right'))).toBeNotVisible();
+    await expect(element(by.text('Left'))).toBeVisible();
+  });
+  it('Should handle scroll', async () => {
+    try {
```
Comment (rotemmiz) [inline detox/test/e2e/p-nested-scroll-views.js]: What are you catching here?

**Comment by rotemmiz** [comment]
Ping ;)

**Comment by heuism** [comment]
@Protothor does this work that @rotemmiz can merge in?

**Comment by timea** [comment]
Can we please get this PR in? Would be very helpful.
